internal/envoy/auth.go

```diff
@@ -47,12 +47,34 @@ var (
 // UpstreamTLSContext creates an auth.UpstreamTlsContext. By default
 // UpstreamTLSContext returns a HTTP/1.1 TLS enabled context. A list of
 // additional ALPN protocols can be provided.
-func UpstreamTLSContext(alpnProtocols ...string) *auth.UpstreamTlsContext {
-	return &auth.UpstreamTlsContext{
+func UpstreamTLSContext(cert []byte, subjaltname []string, alpnProtocols ...string) *auth.UpstreamTlsContext {
+	context := &auth.UpstreamTlsContext{
 		CommonTlsContext: &auth.CommonTlsContext{
 			AlpnProtocols: alpnProtocols,
 		},
 	}
+
+	validation := validationContext(cert, subjaltname)
+	if validation != nil {
+		context.CommonTlsContext.ValidationContextType = validation
+	}
+	return context
+}
+
+func validationContext(cert []byte, subjaltname []string) *auth.CommonTlsContext_ValidationContext {
```

Comment:
two things

1. please put the guard clause first; ie

```go
if cert == nil {
   // no validation required
   return nil
```

b. please don't push this logic into validationContext. If its not supposed to be called with cert == nil, then don't call it with cert == nil. The reason I ask this is no validationContext _sometimes_ returns nil depending on its inputs, that means we have nils floating around the data structures, this is how you get java :). We should at all times be following these two rules

1. Don't use nil to indicate a failure
2. Reduce the propagation of nils through the data structures.